udf_sbm_template.c

```diff
@@ -0,0 +1,408 @@
+                PRF_CSEND_REAL_N(node_host, face_areas, n_faces, compute_node, ND_ND);
+                PRF_CSEND_INT_N(node_host, face_ids, n_faces, compute_node, mnpf);
+
+                RELEASE_MEMORY_N(face_coords, ND_ND);
+                RELEASE_MEMORY_N(face_areas, ND_ND);
+                RELEASE_MEMORY_N(face_ids, mnpf);
+            }
+        }
+#endif /* RP_NODE */
+
+#if RP_HOST
+        compute_node_loop(compute_node) {
+            PRF_CRECV_INT(node_zero, &n_faces, 1, compute_node);
+
+            ASSIGN_MEMORY_N(face_coords, n_faces, real, ND_ND);
+            ASSIGN_MEMORY_N(face_areas, n_faces, real, ND_ND);
+            ASSIGN_MEMORY_N(face_ids, n_faces, int, mnpf);
+
+            PRF_CRECV_REAL_N(node_zero, face_coords, n_faces, compute_node, ND_ND);
+            PRF_CRECV_REAL_N(node_zero, face_areas, n_faces, compute_node, ND_ND);
+            PRF_CRECV_INT_N(node_zero, face_ids, n_faces, compute_node, mnpf);
+
+            for (i_f = 0; i_f < n_faces; i_f++) {
+                for (d = 0; d < ND_ND; d++) {
+                    fprintf(file_faces, "%27.17e ", face_coords[d][i_f]);
+				}
+                for (d = 0; d < ND_ND; d++) {
+                    fprintf(file_faces, "%27.17e ", face_areas[d][i_f]);
+                }
+                for (d = 0; d < mnpf; d++) {
+                    fprintf(file_faces, " %10d", face_ids[d][i_f]);
+                }
+                fprintf(file_faces, "\n");
+            }
+
+            RELEASE_MEMORY_N(face_coords, ND_ND);
+            RELEASE_MEMORY_N(face_areas, ND_ND);
+            RELEASE_MEMORY_N(face_ids, mnpf);
+        } /* close compute_node_loop */
+        fclose(file_faces);
+#endif /* RP_HOST */
+
+    } /* close loop over threads */
+
+    if (myid == 0) {printf("\nFinished UDF store_faces_normals_id.\n"); fflush(stdout);}
+}
+
+  /*-------------------*/
+ /* read_vof_face_ids */
+/*-------------------*/
+
+DEFINE_ON_DEMAND(read_vof_face_ids) {
+
+    if (myid == 0) {printf("\nStarted UDF read_vof_face_ids.\n"); fflush(stdout);}
+
+#if RP_NODE
+    int compute_node;
+#endif /*RP_NODE*/
+
+#if RP_HOST
+    int i, j;
+    char file_ids[] = "face_ids.dat";
+    char file_vof[30];
+    sprintf(file_vof, "inlet_VOFw_start%i.dat", time_step_start);
+    FILE *fp_ids;
+    FILE *fp_vof;
+
+    if (NULLP(fp_ids = fopen(file_ids, "r"))) {
+        Error("\nUDF-error: Unable to open %s for reading\n", file_ids);
+        exit(1);
+    }
+    fscanf(fp_ids, "# %i", &n_faces);
+
+    if (NULLP(fp_vof = fopen(file_vof, "r"))) {
+        Error("\nUDF-error: Unable to open %s for reading\n", file_vof);
+        exit(1);
+    }
+
+    ASSIGN_MEMORY_N(ids, n_faces, int, mnpf);
+    ASSIGN_MEMORY_N(vof_w, n_faces, int, n_time_steps + 1);
+
+    for (i=0; i<n_faces; i++){
+        for (j=0; j<mnpf; j++){
+            fscanf(fp_ids, "%i", &ids[j][i]);
+        }
+        for (j=0; j<n_time_steps + 1; j++){ /* one more than n_time_steps, because initial value is included */
```

Comment:
Isn't it better to increase n_time_step by 1 at begin, instead of adding 1 at lot of places? Risky when adapting code later...

Reply:
It has been adjusted, now it is changed in `read_inlet_fluent.py` before substituting in the udf. I still wanted to keep it hidden for the user, because it is not intuitive to add 1 to the number of time steps.